### ISSUE TITLE
Uniformize `randomwalk!` in gridspaces

### ISSUE THREAD

**Issue by Tortar** [JuliaDynamics/Agents.jl#799]
Currently the implementation of `randomwalk!` for gridspaces is like so:

```julia
function randomwalk!(
    agent::AbstractAgent,
    model::ABM{<:AbstractGridSpace},
    r::Real = 1;
    kwargs...
)
    if abmspace(model).metric == :euclidean
        throw(ArgumentError(
            "Random walks on a `GridSpace` with Euclidean metric are not defined. " *
            "You might want to use a `ContinuousSpace` or a different metric."
        ))
    end
    offsets = offsets_at_radius(model, r)
    walk!(agent, rand(abmrng(model), offsets), model; kwargs...)
end

function randomwalk!(
    agent::AbstractAgent,
    model::ABM{<:GridSpaceSingle},
    r::Real = 1;
    kwargs...
)
    if abmspace(model).metric == :euclidean
        throw(ArgumentError(
            "Random walks on a `GridSpace` with Euclidean metric are not defined. " *
            "You might want to use a `ContinuousSpace` or a different metric."
        ))
    end
    offsets = offsets_at_radius(model, r)
    target_positions = map(β -> normalize_position(agent.pos .+ β, model), offsets)
    idx = findall(pos -> id_in_position(pos, model) == 0, target_positions)
    available_offsets = @view offsets[idx]
    if isempty(available_offsets)
        # don't move, return agent only for type stability
        agent
    else
        walk!(agent, rand(abmrng(model), available_offsets), model; ifempty=false)
    end
end
```

which is odd since the first `randomwalk!` calls `walk!(agent, rand(abmrng(model), offsets), model; kwargs...)`, so that if `ifempty=true` is passed we can have that the agent actually **don't move** if the random position is occupied, in the case of a `SingleGridSpace` instead we always move the agent (provided that there is an unoccupied position), I don't think the docstrings suggests this difference in behaviour between the two, and also having two different behaviour for essentially the same space would be a good thing, so I propose to change the second function in

```julia
function randomwalk!(
    agent::AbstractAgent,
    model::ABM{<:GridSpaceSingle},
    r::Real = 1;
    kwargs...
)
    if abmspace(model).metric == :euclidean
        throw(ArgumentError(
            "Random walks on a `GridSpace` with Euclidean metric are not defined. " *
            "You might want to use a `ContinuousSpace` or a different metric."
        ))
    end
    offsets = offsets_at_radius(model, r)
    walk!(agent, rand(abmrng(model), offsets), model; ifempty=true)
end
```

or just entirely removing the second function and change `kwargs...` in  `ifempty=true` in the first, as the `walk!` function does.

I'm also thinking that maybe we could provide both functions under slightly different names or using some attribute, since even if they have different behaviours, each could be maybe useful...

**Comment by Tortar** [comment]
@Datseris what is your opinion about this? 

Also looking at the ContinuousSpace version, we currently only support the random walk in 2/3 dimensions, we could extend (or change) the code with one of the general approaches described e.g. here https://math.stackexchange.com/questions/1585975/how-to-generate-random-points-on-a-sphere

**Comment by Datseris** [comment]
oh yeah that's a mistake, I guess we should uniformize these two. It is clearly more performant to randomly pick a position and then check if it is empty, rather than checking whetehr there is _any_ empty position and then moving. So all functions should do what the first does.

---

Yes for the random walk in any dimensions, but that seems like it should be a different PR.

**Comment by mastrof** [comment]
Chiming in because I'm interested in following the developments here. I'll take the chance to stress that the current code for the ContinuousSpace random walks does a bit more than taking random points on a sphere. 
Selecting random points on a sphere is nice and fast but prevents you from generating correlated walks. But it's true the current approach is not optimal if one just wants an uncorrelated walk, and it has no immediate N-dimensional generalization (as far as I know).

If you plan to move the current algorithm out and stick with the isotropic reorientations, I could think about developing a small distinct module with these types of extensions to the simple random walk.

**Comment by Tortar** [comment]
I think that maybe the best way is to keep the current code for generating correlated walks (happy to know now what all of that extra code was all about :D) but use a simpler algorithm for generic uncorrelated ones in a N-dimensional space. Let me know if you think it's sound @mastrof , but anyway a small module for simple/correlated random walks seems cool to me anyway if nothing like this exists in Julia

**Comment by Tortar** [comment]
closed by #802
